호텔 조회 기능 구현

nasaoreo opened the pull request:
## 작업 내용
- 호텔, 호텔 방 Entity 및 Repository 추가
- 호텔 Service 및 Controller 추가
- 호텔 Service 및 Controller에 호텔 조회 기능 추가
- 호텔 조회 관련 DTO 추가

## 2차 스프린트 때 보완할 사항
- 예외 핸들링 코드(현재는 Service 로직 수행 시 문제되는 부분에 대해서만 최소한으로 구현)
- VO, Entity Validation 코드
- Controller의 반환 타입 변경(ApiResponse) 

close #10

nasaoreo committed:
Feat: 호텔 조회 기능 구현

- 호텔, 호텔 방 Entity 및 Repository 추가
- 호텔 Service 및 Controller 추가
- 호텔 Service 및 Controller에 호텔 조회 기능 추가
- 호텔 조회 관련 DTO 추가

taehyunnn commented on src/main/java/com/grepp/carrierroute/hotel/controller/HotelController.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    @GetMapping("/hotels/{hotelId}/rooms/{roomId}")
+    public HotelSearchResponseDto getHotelRoom(@PathVariable Long roomId) throws HotelInfoNotFoundedException {
```
```suggestion
    public HotelSearchResponseDto getHotelRoom(@PathVariable Long hotelId, @PathVariable Long roomId) throws HotelInfoNotFoundedException {
```
hotelId가 빠졌습니당~

nasaoreo replied:
> hotelId가 빠졌습니당~

이 부분은 사실 제가 일부러 빼긴 했는데, controller 메소드 호출할 때 hotelId를 사용하지 않거든요!
그래서 URI를 바꾸는 것이 어떨까 고민도 했습니다!ㅜㅜ

taehyunnn commented on src/main/java/com/grepp/carrierroute/hotel/domain/HotelRoom.java:
```diff
@@ -0,0 +1,69 @@
+    {
+        if(count == 0 || maxGuestNumber == 0 || pricePerDay == 0){
+            throw new InvalidHotelRoomParameterException(ErrorMessage.INVALID_HOTEL_ROOM_PARAMETER);
+        }
```
메소드로 분리하면 좋을 것 같네용~

nasaoreo replied:
> 메소드로 분리하면 좋을 것 같네용~

validation 코드 추가할 때 메소드로 분리할 예정입니다!ㅎㅎ

taehyunnn commented on src/main/java/com/grepp/carrierroute/hotel/dto/HotelSearchRequestDto.java:
```diff
@@ -0,0 +1,39 @@
+    {
+        if(guestNumber == 0 || numOfRoom == 0 || !endDate.isAfter(startDate)){
+            throw new InvalidHotelSearchParameterException(ErrorMessage.INVALID_HOTEL_SEARCH_PARAMTER);
+        }
```
이부분도 메소드 분리하면 좋을 것 같아용~

taehyunnn commented on src/main/java/com/grepp/carrierroute/hotel/exception/HotelInfoNotFoundedException.java:
```diff
@@ -0,0 +1,7 @@
+    public HotelInfoNotFoundedException(ErrorMessage errorMessage) {
+        super(errorMessage.getMessage());
+    }
+}
```
Unchecked Exception 이라면 RuntimeException을 상속 받는건 어떨까요?

nasaoreo replied:
> Unchecked Exception 이라면 RuntimeException을 상속 받는건 어떨까요?

좋은 의견이네요! 반영하겠습니당!ㅎㅎ

taehyunnn commented on src/main/java/com/grepp/carrierroute/hotel/service/HotelService.java:
```diff
@@ -0,0 +1,77 @@
+    private final HotelRepository hotelRepository;
+    private final HotelConverter converter;
+
+    public HotelSearchResponseDto findRoomBy(Long id) throws HotelInfoNotFoundedException {
+        return hotelRoomRepository.findById(id)
+                .map(converter::convertHotelSearchResponseDto)
+                .orElseThrow(()->new HotelInfoNotFoundedException(ErrorMessage.HOTEL_ROOM_NOT_FOUNDED));
+    }
```
조회만 할 때는 @Transactional(readOnly = true) 로 해주면 좋을 것 같아요~

taehyunnn reviewed (commented):
👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 
의견 몇 개 남기고 갑니다 ㅎ 내일 스크럼때 만나용~~

dia312 commented on src/main/java/com/grepp/carrierroute/hotel/domain/HotelRoom.java:
```diff
@@ -0,0 +1,69 @@
+                     long pricePerDay,
+                     String photoUrl,
+                     @NonNull Hotel hotel) throws InvalidHotelRoomParameterException
+    {
```
count, maxGuestNumber, pricePerDay 도 @NonNull 어노테이션 붙이면 어떨까요?

nasaoreo replied:
> count, maxGuestNumber, pricePerDay 도 @nonnull 어노테이션 붙이면 어떨까요?

primitive 타입을 사용했기 때문에 @nonnull 어노테이션을 붙이는 것이 의미가 없어 validation 코드를 추가한 것인데요! 이 부분때문에 Wrapper 타입을 사용할까 고민 중입니다. 스크럼 시간에 같이 얘기해봐용!ㅎㅎ

dia312 reviewed (commented):
줄리~~~! 줄리랑 헨리 코드 보면서 많이 배우고 이번 이슈 구현하는데 많이 도움된 것 같아요!! 앞으로도 화이팅하겠습니다!! 💯

nasaoreo committed:
Refactor: 코드 리뷰를 반영하여 호텔 조회 코드 리팩토링

- Entity, DTO에서 사용된 원시 타입을 Wrapper 타입으로 변경
- Entity, DTO의 예외 핸들링 코드를 별도의 메소드로 분리
- Service의 @Transactional 어노테이션 옵션 추가(readOnly = true)
- 예외 클래스가 RuntimeException 상속받도록 변경

dia312 reviewed (approved):
확인했습니다~!!! 👍
